### PR TITLE
SDL 0064 Choice VR Optional

### LINKED ISSUE #943: [SDL 0064] Choice-VR optional
Proposal: [Choice-VR optional](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0064-choice-vr-optional.md)

> This proposal is about changing Choice.vrCommands to be not mandatory.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/189

Steering Committee Decision:

> The Steering Committee agreed to accept this proposal and also agreed that the error for mixed vrCommand should be more descriptive, like info = "Some choices don't contain VR commands. Either all or none must have voice commands."

### PR DESCRIPTION
**Pull request by NicoleYarroch**
Fixes #943 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Test cases added.

### Summary
The `Choice` RPC's `vrCommands` are no longer mandatory.

##### Enhancements
* The `Choice` RPC `vrCommands` can now be set to `nil`.
* The screen manager now makes sure that either all or none of the choiceset items have unique `vrCommands`.

### Tasks Remaining:
- [x] Test with Core (v.5.0)
- [x] Test with Manticore (v.4.5.1)

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLChoiceSet.m [hunk 1]
```diff
@@ -62,23 +62,31 @@ - (instancetype)initWithTitle:(NSString *)title delegate:(id<SDLChoiceSetDelegat
...
+    NSMutableSet *vrCheckSet = [NSMutableSet set];
     for (SDLChoiceCell *cell in choices) {
         [choiceTextSet addObject:cell.text];
+        if (cell.voiceCommands.count > 0) {
+            [vrCheckSet addObject:cell.voiceCommands];
+        }
     }
     if (choiceTextSet.count < choices.count) {
-        SDLLogW(@"Attempted to create a choice set with duplicate cell text. Cell text must be unique.");
+        SDLLogW(@"Attempted to create a choice set with duplicate cell text. Cell text must be unique. The choice set will not be set.");
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLChoiceSet.m]: Fixed

File: SmartDeviceLink/SDLMenuManager.m [hunk 2]
```diff
@@ -117,16 +117,27 @@ - (void)setMenuCells:(NSArray<SDLMenuCell *> *)menuCells {
...
+
+    // Check for duplicate titles
     if (titleCheckSet.count != menuCells.count) {
         SDLLogE(@"Not all cell titles are unique. The menu will not be set.");
         return;
     }
 
+    // Check for duplicate voice recognition commands
+    NSMutableSet<NSArray<NSString *> *> *choiceVoiceCommandSet = [NSMutableSet setWithArray:voiceCommandCheckArray];
+    if (choiceVoiceCommandSet.count < voiceCommandCheckArray.count) {
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLMenuManager.m]: Fixed. I also refactored the duplicate VR check in the `SDLChoiceSet` class.

File: SmartDeviceLink/SDLChoiceSet.m [hunk 3]
```diff
@@ -62,23 +62,29 @@ - (instancetype)initWithTitle:(NSString *)title delegate:(id<SDLChoiceSetDelegat
...
     }
 
-    NSMutableArray<NSArray<NSString *> *> *nonNilVoiceCommands = [NSMutableArray array];
-    for (SDLChoiceCell *cell in choices) {
-        if (cell.voiceCommands != nil) {
-            [nonNilVoiceCommands addObject:cell.voiceCommands];
-        }
+    // All or none of the choices must have VR commands
+    if ((choiceSetWithVoiceCommandCount > 0 && choiceSetWithVoiceCommandCount < choices.count)) {
+        SDLLogE(@"If using voice recognition commands, all of the choice set cells must have unique VR commands. There are %lu unique VR commands and %lu choice set items. The choice set will not be set.", (unsigned long)choiceSetWithVoiceCommandCount, (unsigned long)choices.count);
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLChoiceSet.m]: "...There are %lu cells with unique voice commands and %lu total cells"

File: SmartDeviceLink/SDLChoiceSet.m [hunk 4]
```diff
@@ -62,23 +62,29 @@ - (instancetype)initWithTitle:(NSString *)title delegate:(id<SDLChoiceSetDelegat
     }
 
     NSMutableSet<NSString *> *choiceTextSet = [NSMutableSet setWithCapacity:choices.count];
+    NSMutableSet<NSString *> *allMenuVoiceCommands = [NSMutableSet set];
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLChoiceSet.m]: Fixed

File: SmartDeviceLink/SDLChoiceSet.m [hunk 5]
```diff
@@ -62,23 +62,29 @@ - (instancetype)initWithTitle:(NSString *)title delegate:(id<SDLChoiceSetDelegat
...
+    NSUInteger allVoiceCommandsCount = 0;
+    NSUInteger choiceSetWithVoiceCommandCount = 0;
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLChoiceSet.m]: Fixed

File: SmartDeviceLink/SDLChoiceSet.m (hunk 3 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLChoiceSet.m]: Fixed

**Pull request merged**